data_steward/cdr_cleaner/cleaning_rules/deid/conflicting_hpo_state_generalization.py

```diff
@@ -0,0 +1,258 @@
+"""
+Cleaning rule to generalize conflicting HPO states
+"""
+import logging
+import os
+
+from google.cloud import bigquery
+from google.api_core.exceptions import Conflict
+
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule, query_spec_list
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+from common import JINJA_ENV, OBSERVATION
+from resources import DEID_PATH
+
+LOGGER = logging.getLogger(__name__)
+
+JIRA_ISSUE_NUMBERS = ['DC512', 'DC834']
+JIRA_ISSUE_URL = [
+    'https://precisionmedicineinitiative.atlassian.net/browse/DC-512',
+    'https://precisionmedicineinitiative.atlassian.net/browse/DC-834'
+]
+
+MAP_TABLE_NAME = "_mapping_person_src_hpos"
+MAP_HPO_ALLOWED_STATES = "_mapping_src_hpos_to_allowed_states"
+UNIT_MAPPING_TABLE_DISPOSITION = "WRITE_TRUNCATE"
+
+HPO_ID_NOT_RDR_QUERY = JINJA_ENV.from_string("""
+  SELECT
+  DISTINCT person_id, src_hpo_id
+  FROM
+    `{{dataset_id}}._mapping_{{table}}`
+  JOIN
+    `{{dataset_id}}.{{table}}`
+  USING
+    `{{table}}_id`
+  WHERE
+    src_hpo_id NOT LIKE 'rdr'
+""")
+
+LIST_PERSON_ID_TABLES = JINJA_ENV.from_string("""
+  SELECT table_name
+  FROM `{{project_id}}.{{dataset_id}}.INFORMATION_SCHEMA.COLUMNS`
+  WHERE lower(column_name) = 'person_id'
+""")
+
+INSERT_TO_MAP_TABLE_NAME = JINJA_ENV.from_string("""
+  INSERT INTO `{{project_id}}.{{sandbox_dataset_id}}.{{table_name}}`
+  (person_id,
+   src_hpo_id)
+  {{select_query}}
+""")
+
+SANDBOX_QUERY_TO_FIND_RECORDS = JINJA_ENV.from_string("""
+  CREATE OR REPLACE TABLE
+  `{{project_id}}.{{sandbox_dataset_id}}.{{target_table}}` AS (
+  SELECT observation_id FROM (
+    SELECT DISTINCT src_hpo_id, obs.person_id, value_source_concept_id, observation_id
+    FROM {{project_id}}.{{sandbox_dataset_id}}.{{map_table_name}} AS person_hpos
+    LEFT JOIN {{project_id}}.{{sandbox_dataset_id}}.{{map_hpo_allowed_states}}
+    USING (src_hpo_id, value_source_concept_id)
```

Comment:
This throws an error once the integration test gets far enough to execute it.

Reply:
I'm attempting to split the query from 'config.json'

"observation_id IN (SELECT observation_id FROM ( ",
                    "SELECT DISTINCT src_hpo_id, obs.person_id, value_source_concept_id, observation_id ",
                    "FROM :idataset._mapping_person_src_hpos AS person_hpos ",
                    "JOIN :idataset.observation AS obs ",
                    "USING (person_id) ",
                    "LEFT JOIN :idataset._mapping_src_hpos_to_allowed_states ",
                    "USING (src_hpo_id, value_source_concept_id) ",
                    "WHERE observation_source_concept_id = 1585249  AND state IS NULL))"

into a Sandbox query and Dataset query.

But at this point, since the Sandbox does not have 'observation' table, I had to skip the 'JOIN :idataset.observation USING (person_id)'.
I was expecting to find some error in they query at the time of running the Integration test. Just out of curiosity, what error di you see? I looked up 'src_hpos_to_allowed_state.csv' file, based on that and '_mapping_person_src_hpos' data, the LEFT JOIN cannot happen USING both src_hpo_id, value_source_concept_id.

What could be a solution to this? Do you think if I just do the 'LEFT JOIN USING (src_hpo_id)'?

Reply:
@ratuagga I put some info on DC-834.  If you want, we can talk through it.

Reply:
Yes, I have asked a couple of questions in the ticket comments based on the description you gave.